sdk/agentserver/azure-ai-agentserver-core/azure/ai/agentserver/core/client/tools/_client.py

```diff
@@ -0,0 +1,224 @@
+				self._config.retry_policy,
+				self._config.authentication_policy,
+				self._config.custom_hook_policy,
+				self._config.logging_policy,
+				policies.DistributedTracingPolicy(**kwargs),
+				policies.SensitiveHeaderCleanupPolicy(**kwargs) if self._config.redirect_policy else None,
+				self._config.http_logging_policy,
+			]
+		self._client: PipelineClient = PipelineClient(base_url=endpoint, policies=_policies, **kwargs)
+	
+		# Initialize specialized clients with client and config
+		self._mcp_tools = MCPToolsOperations(client=self._client, config=self._config)
+		self._remote_tools = RemoteToolsOperations(client=self._client, config=self._config)
+	
+	def list_tools(self) -> List[FoundryTool]:
+		"""List all available tools from configured sources.
+		
+		Retrieves tools from both MCP servers and Azure AI Tools API endpoints,
+		returning them as FoundryTool instances ready for invocation.
+		:return: List of available tools from all configured sources.
+		:rtype: List[~AzureAITool]
+		:raises ~exceptions.OAuthConsentRequiredError:
+			Raised when the service requires user OAuth consent.
+		:raises ~exceptions.MCPToolApprovalRequiredError:
+			Raised when tool access requires human approval.
+		:raises ~azure.core.exceptions.HttpResponseError:
+			Raised for HTTP communication failures.
+
+		"""
+
+		existing_names: set[str] = set()
+
+		tools: List[FoundryTool] = []
+
+		# Fetch MCP tools
+		mcp_tools = self._mcp_tools.list_tools(existing_names)
+		tools.extend(mcp_tools)
+
+		# Fetch Tools API tools
+		tools_api_tools = self._remote_tools.resolve_tools(existing_names)
+		tools.extend(tools_api_tools)
+
+		for tool in tools:
+			# Capture tool in a closure to avoid shadowing issues
+			def make_invoker(captured_tool):
+				return lambda *args, **kwargs: self.invoke_tool(captured_tool, *args, **kwargs)
+			tool.invoker = make_invoker(tool)
+		return tools
+	
+	def invoke_tool(
+		self,
+		tool: Union[str, FoundryTool],
+		*args: Any,
+		**kwargs: Any,
+	) -> Any:
+		"""Invoke a tool by instance, name, or descriptor.
+		
+		:param tool: Tool to invoke, specified as an AzureAITool instance,
+			tool name string, or FoundryTool.
+		:type tool: Union[str, ~FoundryTool]
+		:param args: Positional arguments to pass to the tool
+		"""
+		descriptor = self._resolve_tool_descriptor(tool)
+		payload = InvocationPayloadBuilder.build_payload(args, kwargs, {})
+		return self._invoke_tool(descriptor, payload, **kwargs)
+	
+	def _resolve_tool_descriptor(
+		self, tool: Union[AzureAITool, str, FoundryTool]
+	) -> FoundryTool:
+		"""Resolve a tool reference to a descriptor."""
+		if isinstance(tool, AzureAITool):
+			return tool._descriptor
+		if isinstance(tool, FoundryTool):
+			return tool
+		if isinstance(tool, str):
+			# Fetch all tools and find matching descriptor
+			descriptors = self.list_tools()
+			for descriptor in descriptors:
+				if descriptor.name == tool or descriptor.key == tool:
+					return descriptor
```

Comment:
How do you know that this will be type FoundryTool? If its registered name/key then Foundry has a matching tool?